src/implementation/compression_test.cpp

```diff
@@ -0,0 +1,58 @@
+#include "gtest/gtest.h"
+#include "compression.cpp"
+
+vector<ll> populateVector(int n) {
+    vector<ll> v;
+    rep(i,0,n) {
+        v.push_back(rand() - RAND_MAX/2);
+    }
+    return v;
+}
+
+TEST(COMPRESSION, ManualCase) {
+    vector<ll> v = {-3, 8, 2, 4, 4, 2, 4, -10};
+    compress(v);
+    EXPECT_EQ(v, vector<ll>({1, 4, 2, 3, 3, 2, 3, 0}));
+}
+
+void check(vector<ll>& v) {
+    vector<ll> v0 = v;
+    compress(v);
+    vector<bool> seen(v.size(), false);
+    for(auto x : v) {
+        EXPECT_LE(0, x);
+        EXPECT_LT(x, v.size());
+        seen[x] = true;
+    }
+    rep(i,0,v.size()-1) {
+        if(!seen[i]) {
+            EXPECT_EQ(seen[i], seen[i+1]);
+        }
+    }
+    rep(i,0,v.size()) rep(j,0,v.size()) {
+        EXPECT_EQ(v[i] < v[j], v0[i] < v0[j]);
+    }
+}
+
+TEST(COMPRESSION, Works) {
+    srand(123);
+    rep(i,0,100) {
+        const int SZ = 100;
+        auto v = populateVector(SZ);
+        check(v);
+    }
+}
+
+TEST(COMPRESSION, WorksDuplicate) {
+    srand(123);
+    rep(i,0,100) {
+        const int SZ = 50;
+        auto v = populateVector(SZ);
+        rep(i,0,SZ) {
+            int times = rand() % 5;
+            rep(j,0,times) v.push_back(v[i]);
+        }
+        random_shuffle(v.begin(), v.end());
```

Comment:
Beware that `srand` has no effect here, because `random_shuffle` (and AFAIK everything in std) doesn't rely on `rand`.

Reply:
Just tested and it seems it does indeed have an effect. Compiler: `g++ (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4`

Code:
```c++
vector<int> v = {1,2,3};
random_shuffle(v.begin(), v.end());
cout << v << endl; // 1 3 2
```
vs
```c++
vector<int> v = {1,2,3};
srand(2);
random_shuffle(v.begin(), v.end());
cout << v << endl; // 2 3 1
```
Interestingly, `random_shuffle` seems to have a fixed seed even when I don't use `srand` - it always produces the same result, so hopefully even in an implementation without `rand` it should be deterministic. Not that it matters too much.

Reply:
Not using `srand` should equal to calling `srand(1)` according to the specs.

The behavior of `random_shuffle` has surprised me. I brought it up in the first place only because I've already dealt with a similar quirk once. But now I recall the offending code was explicitly creating a `random_device`, so maybe that is the reason this code works as expected. By the way, the documentation on `random_shuffle` says the behavior is implementation-defined, but `rand` is often used.

Anyway, as you say, this is no big deal.